File: Sliding Window Maximum/Sliding_Window_Maximum.cpp
// 按照题目的要求，是线性的时间复杂度

// Runtime: 56 ms, faster than 94.55% of C++ online submissions for Sliding Window Maximum.
// Memory Usage: 13.2 MB, less than 44.28% of C++ online submissions for Sliding Window Maximum.

class Solution 
{
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) 
    {
        vector<int> res;

        // 边界条件处理
        if (nums.size() == 0)    
            return res;
        
        deque<int> memo;

        for (int i = 0; i < k; i++)
        {
            while (!memo.empty() && nums[i] > nums[memo.back()])
                memo.pop_back();
            memo.push_back(i);
        }

        for (int i = k; i < nums.size(); i++)
        {
            res.push_back(nums[memo.front()]);

            while (!memo.empty() && nums[i] > nums[memo.back()])
                memo.pop_back();
            memo.push_back(i);

            if (i - memo.front() >= k)
                memo.pop_front();
        }

        res.push_back(nums[memo.front()]);
        return res;
    }
};

File: Sliding Window Maximum/Sliding_Window_Maximum.py
# Runtime: 116 ms, faster than 91.48% of Python3 online submissions for Sliding Window Maximum.
# Memory Usage: 20 MB, less than 23.66% of Python3 online submissions for Sliding Window Maximum.

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        res = []

        if len(nums) is 0:
            return res
        
        memo = []

        for i in range(k):
            while memo and nums[i] > nums[memo[-1]]:
                memo.pop()
            memo.append(i)
        
        for i in range(k, len(nums)):
            res.append(nums[memo[0]])

            while memo and nums[i] > nums[memo[-1]]:
                memo.pop()
            memo.append(i)

            if i - memo[0] >= k:
                memo.pop(0)
        
        res.append(nums[memo[0]])
        return res